catalogue_graph/src/ingestor_loader.py

```diff
@@ -28,37 +32,202 @@ class IngestorLoaderConfig(BaseModel):
     is_local: bool = False
 
 
-def extract_data(start_offset: int, end_index: int, is_local: bool) -> list[dict]:
-    print("Extracting data from Neptune ...")
-    client = get_neptune_client(is_local)
+# Maximum number of related nodes to return for each relationship type
+RELATED_TO_LIMIT = 10
 
-    limit = end_index - start_offset
-    print(f"Processing records from {start_offset} to {end_index} ({limit} records)")
+# There are a few Wikidata supernodes which cause performance issues in queries.
+# We need to filter them out when running queries to get related nodes.
+# Q5 -> 'human', Q151885 -> 'concept'
+IGNORED_WIKIDATA_IDS = ["Q5", "Q151885"]
+
+
+def get_related_query(
+    edge_type: str,
+    direction: str = "from",
+    source_concept_label_types: list[str] | None = None,
+) -> str:
+    """
+    Return a parameterized Neptune query to fetch related Wellcome concepts:
+        1. For each Wellcome concept (`Concept` node), retrieve its associated `SourceConcept` nodes.
+        2. For each `SourceConcept`, traverse edges of type `edge_type` in the specified `direction`
+        (e.g. `edge_type="NARROWER_THAN"` combined with `direction="from"` would yield broader concepts)
+        to find related `SourceConcept` nodes.
+        3. Get the Wellcome concept(s) associated with each related `SourceConcept`, deduplicate with a `WITH`
+        clause and return the most popular concepts (determined by the number of Works in which they appear).
+    """
+    label_filter = ""
+    if source_concept_label_types is not None and len(source_concept_label_types) > 0:
+        label_filter = "WHERE " + " OR ".join(
+            [f"linked_source_concept:{c}" for c in source_concept_label_types]
+        )
+
+    left_arrow = "<" if direction == "to" else ""
+    right_arrow = ">" if direction == "from" else ""
+
+    return f"""
+        MATCH (concept:Concept)
```

Comment:
It might be useful to break down this query a bit with comments as although the comment at the start of this function explains what's going in it's still quite hard to parse mentally.